README.md

```diff
@@ -714,77 +862,73 @@ where they are needed.
-could hold a reference to its `stringref` directly.  There would then be
-no need for cursor validity checks.  On the other hand, then we would
-have a new type that would infiltrate everything, from implementation to
-JavaScript API to the type system and so on.  And, absent compiler
-heroics, reference-typed cursors may cause high allocation overheads.
-
-The other way you could make cursors opaque would be as opaque scalar
-values.  The idea is that a cursor is really an `i32` under the hood,
-but its value isn't accessible.  Such a cursor wouldn't stand alone in
-the way reference-typed cursors would: you need to pass a `stringref`
-and a cursor to instructions, and you need to check the cursor for
-validity with regards to the string.  We still have some of the type
-profusion issues from a "cognitive load" point of view.  But, you
-couldn't observe the difference in cursor values between
-implementations, which would be a nice property.
-
-In the end though, besides simplicity, what tipped the balance towards
-plain `i32` values was precisely that string cursors could be meaningful
-to the host instead of opaque.  A host should be able to reason about
-string positions and communicate those positions to WebAssembly -- after
-all, `stringref` values belong to the host too.  Specifying that string
-cursors are code unit offsets makes this possible, while also
-constraining e.g. WebAssembly implementations in different web browsers
-to all use the same notion of string offsets.
-
-See https://github.com/wingo/wasm-strings/issues/6 and
-https://github.com/wingo/wasm-strings/issues/11 for a full discussion.
+We expect that web browsers use JS strings as `stringref`.  WTF-16
+access uses the same primitives exposed to JavaScript.  Iterating the
+codepoints of a string using the WTF-8 interfaces might use a small
+per-instance cache that associates the most recent WTF-16 and WTF-8
+offsets for the last-accessed string or strings.
+
+There is a possibility that some web browsers may eventually switch from
+the one-byte/two-byte representation to WTF-8 with breadcrumbs, which
+would make browsers follow the same strategy as the non-browser case.
+
+Without a cache, accessing each codepoint in a string via
+`string.get_wtf8` would be quadratic.  Perhaps this proposal should
+specify time complexity bounds for some kinds of string algorithm, for
+example visiting each codepoint in a string.
+
+### What are the expected performance characteristics of the WTF-8 and WTF-16 interfaces?
+
+We expect that compilers that emit WTF-8 instructions prefer chunks of
+`string.encode_wtf8` over `string.get_wtf8` / `string.advance_wtf8`.
+`string.get_wtf8` / `string.advance_wtf8` will be faster on WebAssembly
+implementations that use WTF-8 internally.
+
+We expect that compilers that emit the WTF-16 interface place more
+importance on `string.get_wtf16`.  Implementations should ensure that
+`string.get_wtf16` runs in near-linear time, even on systems that
```

Comment:
Do you mean "near-constant"? Linear time for a single `get` would be pretty slow.

Reply:
Hah, right, a thinko on my part -- will fix.